### ISSUE TITLE
Cannot build on CentOS6: C++ compiler is needed

### ISSUE THREAD

**Issue by ndevenish** [wxWidgets/Phoenix#780]
**Operating system**: CentOS6/RHEL6
**wxPython version**: 4.0.1/master
**Python version**: 2.7.14
**Description of the problem**:
Cannot compile on RHEL6, or a CentOS docker image of the same environment. Encountered on a production RHEL6 system via pip, reproduced on a CentOS6 `Dockerfile`:
```Dockerfile
FROM centos:6

RUN yum install -y gcc gcc-c++ make zlib-devel openssl-devel git

# WX dependencies, not required to demonstrate this problem
# RUN yum install -y  freeglut freeglut-devel gstreamer-plugins-base-devel gtk2-devel libjpeg-devel libnotify-devel libSM-devel libtiff-devel libXtst-devel SDL-devel webkitgtk-devel

# Install python 2.7
RUN mkdir /tmp/python && \
    cd /tmp/python && \
    curl -L https://www.python.org/ftp/python/2.7.14/Python-2.7.14.tgz | tar -xz --strip-components=1 && \
    ./configure --prefix=/opt/python && \
    make install
ENV PATH="/opt/python/bin:${PATH}"

# Get the source and build
RUN cd /tmp && git clone --recursive https://github.com/wxWidgets/Phoenix.git
RUN cd /tmp/Phoenix && python build.py --gtk2 build
```

Get the error during configuration of wx:

```
...
checking for g++... g++
checking whether we are using the GNU C++ compiler... no
checking whether g++ accepts -g... yes
checking whether we are using the Intel C++ compiler... no
checking whether we are using the Sun C++ compiler... no
configure: error: C++ compiler is needed to build wxWidgets
Traceback (most recent call last):
  File "build.py", line 1295, in cmd_build_wx
    wxbuild.main(wxDir(), build_options)
  File "/tmp/Phoenix/buildtools/build_wxwidgets.py", line 374, in main
    "Error running configure")
  File "/tmp/Phoenix/buildtools/build_wxwidgets.py", line 85, in exitIfError
    raise builder.BuildError(msg)
BuildError
```

However, calling the configuration that the build prints out directly:
```
$ /tmp/Phoenix/ext/wxWidgets/configure --enable-unicode --with-gtk=2 --with-opengl --enable-sound --enable-graphics_ctx --enable-mediactrl --enable-display --enable-geometry --enable-debug_flag --enable-optimise --disable-debugreport --enable-uiactionsim --with-sdl
```
**works**, so this appears to be a problem in the way Phoenix calls the wx builder, not a problem of wx directly?

**Comment by mesalu** [comment]
Phoenix tries building with C++11.
the stock gcc on centos6 is, surprisingly, too old to support that standard.

**Comment by RobinD42** [comment]
And the reason the separate configure works is we're adding '-std=c++11' to CXXFLAGS so it trying to use the flag when we call wx's configure. Without that configure will succeed, the build will fail when it gets to the wxPython code.

**Comment by pauldmccarthy** [comment]
I've had success compiling wxPython 4.0.1 on CentOS 6.6, simply by replacing `'-std=c++11'` with `''` in `build.py`:

https://git.fmrib.ox.ac.uk/fsl/fsleyes/fsleyes-docker-images/blob/master/scripts/build_install_wxpython4.sh

It seems that there is currently no C++11 code in the sip wrappers that come bundled with the 4.0.1 tarball, so there is no need for a C++11 compiler.

I presume that this is not a long-term solution, but it seems to work fine for me.

(The Dockerfile which calls the above linked script is here: https://git.fmrib.ox.ac.uk/fsl/fsleyes/fsleyes-docker-images/blob/master/fsleyes-centos6/Dockerfile)

**Comment by mesalu** [comment]
Awesome!
I suspect that anyone still relegated to cent OS 6 has to support distribution to multiple nodes.
Does this work for you @ndevenish?

**Comment by RobinD42** [comment]
> I've had success compiling wxPython 4.0.1 on CentOS 6.6, simply by replacing '-std=c++11' with '' in build.py:

Thanks, I had planned on testing that today as I remembered that there were some related changes in the latest sip release that might have (and apparently did) changed the need for the c++11 flag.

**Comment by pauldmccarthy** [comment]
http://pyqt.sourceforge.net/Docs/sip4/incompatibilities.html#sip-v4-19-4

The text is a little vague, but it sounds like SIP will generate code requiring a C++11 compiler _only if_ enums are used. I've never used SIP before, so I may be wrong. 

But wx is compiling and running fine for me :)

**Comment by ndevenish** [comment]
As a workaround manually editing the build looks okay - but was the intention really to drop all non-C++11 platforms? I mean, okay if it is - it's your choice - but the main page still mentioned a minimum VS2008 on windows and something about gcc-4.2 on mac (though wxWidgets says OSX10.7, which I believe does have C++11 support), it's be great if extra restrictions above wxWidget's minimum requirements were mentioned (especially since this particular error is imposing flags on the wxWidgets build that it doesn't need).

Also if it is SIP and strong enums, some earlier platforms might support required features with -std=c++0x rather than -std=c++11? (strong enums appear to have been added in GCC 4.4 ). And that SIP page mentioned also using python's enum module, which makes me suspect that the extra dependency might be for Python3 in any case (you don't seem to require enum34).

I'm the first to admit that CentOS6 is a platform I'd love to not have to support - we already patch a load of wxPython3 problems that are fixed post-last-release, but it's hard to persuade people to move to more than one modern thing at a time...

TL;DR Please don't push this requirement up unless you absolutely have to or it's definitely holding you back

**Comment by mesalu** [comment]
As per usual, @RobinD42 gets the final say on things, but here are my thoughts on this matter:

CentOS6, and I guess the equivalent era of red hat, are probably the only distributions that will suffer from not having c++11 support in the stock compiler. They are *old*, even by enterprise linux standards. Sure, there are other distributions from that time period (circa 2011) such as OS X 10.7, but those have different dependencies and receive updates more regularly. 

I struggle to justify supporting it explicitly in the context of Phoenix. Here's why:
Phoenix's dependency on wxWidgets and SIP puts it in a bit of an awkward spot when it comes to compiler support. A change in wxWidgets can cause a sudden shift in requirements generated by SIP, that still may not be necessary on wxWidget's side. 
This is potentially hazardous for Phoenix, especially now that we're migrating towards wrapping wx's `master` branch, rather than the frozen-ish `WX_3_0_BRANCH`. It could be that changes upstream add and remove conditions that cause c++11 support requirements from SIP. Or SIP could start outputting c++11-only code by default. I suspect (Though I wasn't there / paying attention when the flag was added to our configure call) that one of those two cases is what instigated the c++11 requirement in Phoenix initially. We could drop the requirement now, but it will very likely pop back up unexpectedly, and we'd be here again. 
So to recap a bit, our c++11 requirement is a blanket C.Y.A. measurement, and it alienates so few distributions that it is worthwhile, for me, to keep.

I get having to support antiquated systems (coincidentally enough, I have to support CentOS6 too)
Here's what I recommend, I suggest that you test your theory on using c++0x out with a local build. If it works and appears stable you can bundle it with:
`python build.py bdist_wheel` which will generate a wheel in ./dist (you can do bdist_egg too if you need that). If you don't already have access to an inhouse pypi server you can easily set one up with a multitude of pypi server packages such as [pypiserver](https://pypi.python.org/pypi/pypiserver) which are easy to configure, add packages to, and use (pip's documentation on extra indexes is good enough to get rolling). 
This will give you an approach for using a stable build of Phoenix for your migration. 

Additionally, I've been meaning to set up an auxiliary build environment for @RobinD42. I make no promises for a timeframe, or even if it will ever get done. But I'll look into adding CentOS6 compatible builds. 
This doesn't directly solve your problem though, since standard pypi distributions for linux will still try building locally. So you'll still have to acquire your wheel/egg from [our extra package list](https://wxpython.org/Phoenix/snapshot-builds/linux/).

**Comment by Metallicow** [comment]
I'd like @RobinD42 to write a sort of will as to who might or might not pickup the project at such time he passes.

**Comment by RobinD42** [comment]
Closed by #800
